libraries/backend/helpers.rb

```diff
@@ -0,0 +1,332 @@
+        #   2018-08-01,2018-07-01,2018-06-01,2018-05-01,2018-02-01,2018-01-01,2017-12-01,2017-08-01,2017-06-01,
+        #   2017-05-10,2017-05-01,2017-03-01,2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,
+        #   2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'."
+        #
+        # There are cases where the stable api_versions are too old and don't return JSON response.
+        # If the latest stable is too old (based on the age_criteria), then return the preview versions as well.
+        # This is a quick fix until TODO finding a more stable solution.
+        stable_api_versions = message.scan(/\d{4}-\d{2}-\d{2}[,']/).map(&:chop).sort.reverse
+        preview_api_versions = message.scan(/\d{4}-\d{2}-\d{2}-preview/).sort.reverse
+        if wrong_api_version
+          stable_api_versions.delete(wrong_api_version) if stable_api_versions.include?(wrong_api_version)
+          preview_api_versions.delete(wrong_api_version) if preview_api_versions.include?(wrong_api_version)
+        end
+        age_criteria = 2
+        n_a_l = normalize_api_list(age_criteria, stable_api_versions, preview_api_versions)
+        n_a_l.first
+      end
+    end
+  end
+end
+
+# Decide whether to include preview api-versions in the api_Version list.
+# If the latest stable is too old (based on the age_criteria) and there is a newer preview,
+#   then return the preview versions as well.
+def normalize_api_list(age_criteria, stable_versions, preview_versions)
+  return stable_versions if preview_versions.empty?
+  return preview_versions if stable_versions.empty?
+  r_stable_versions = stable_versions.sort.reverse
+  return_list = r_stable_versions
+  r_preview_versions = preview_versions.sort.reverse
+  latest_stable_year = r_stable_versions.first[0..3].to_i
+  latest_preview_year = r_preview_versions.first[0..3].to_i
+  if latest_stable_year < (Time.now.year - age_criteria) && latest_preview_year > latest_stable_year
+    return_list += r_preview_versions
+  end
+  return_list.sort.reverse
+end
+
+# Create necessary Azure environment variables and provide access to them
+#
+# @example:
+#   my_env = AzureEnvironments.get_endpoint('azure_cloud')
+#   my_env.resource_manager_endpoint_url => 'https://management.azure.com/'
+#
+# For graph api endpoint urls:
+#   https://docs.microsoft.com/en-us/graph/deployments
+# For graph api endpoint api versions:
+#   https://docs.microsoft.com/en-us/azure/active-directory/develop/microsoft-graph-intro
+class AzureEnvironments
+  # Following data can be modified if necessary.
+  ENDPOINTS = {
+    'azure_cloud' => {
+      resource_manager_endpoint_url: MicrosoftRestAzure::AzureEnvironments::AzureCloud.resource_manager_endpoint_url,
+      active_directory_endpoint_url: MicrosoftRestAzure::AzureEnvironments::AzureCloud.active_directory_endpoint_url,
+      resource_manager_endpoint_api_version: '2020-01-01',
```

Comment:
Do we want to pin these? Perhaps there is a way to always use the latest? I feel like this is something that could drift from the latest quite easily.

Reply:
The methods used for resource construction (getting the latest API versions of resource providers, server-side filtering, etc.) rely on data format used by `resource_manager_endpoint`. Changing this API version without testing would potentially break the base class and every resource based on that. 

I will add a reminder to check new API versions when available.